catboost/libs/fstr/shap_values.cpp

```diff
@@ -989,29 +1117,49 @@ TShapPreparedTrees PrepareTrees(
         leafWeights = CollectLeavesStatistics(*dataset, model, localExecutor);
     }
 
-    TShapPreparedTrees preparedTrees;
-    preparedTrees.CalcShapValuesByLeafForAllTrees = IsPrepareTreesCalcShapValues(model, dataset, mode);
+    preparedTrees->CalcShapValuesByLeafForAllTrees = IsPrepareTreesCalcShapValues(model, dataset, mode);
 
-    if (!preparedTrees.CalcShapValuesByLeafForAllTrees) {
+    if (!preparedTrees->CalcShapValuesByLeafForAllTrees) {
         TVector<double> modelLeafWeights(model.ModelTrees->GetLeafWeights().begin(), model.ModelTrees->GetLeafWeights().end());
-        preparedTrees.LeafWeightsForAllTrees
+        preparedTrees->LeafWeightsForAllTrees
             = modelLeafWeights.empty() ? leafWeights : modelLeafWeights;
     }
 
-    preparedTrees.ShapValuesByLeafForAllTrees.resize(treeCount);
-    preparedTrees.SubtreeWeightsForAllTrees.resize(treeCount);
-    preparedTrees.MeanValuesForAllTrees.resize(treeCount);
-    preparedTrees.AverageApproxByTree.resize(treeCount);
-    preparedTrees.CalcInternalValues = calcInternalValues;
+    preparedTrees->ShapValuesByLeafForAllTrees.resize(treeCount);
+    preparedTrees->SubtreeWeightsForAllTrees.resize(treeCount);
+    preparedTrees->MeanValuesForAllTrees.resize(treeCount);
+    preparedTrees->AverageApproxByTree.resize(treeCount);
+    preparedTrees->CalcInternalValues = calcInternalValues;
 
     const TModelTrees& forest = *model.ModelTrees;
     MapBinFeaturesToClasses(
         forest,
-        &preparedTrees.BinFeatureCombinationClass,
-        &preparedTrees.CombinationClassFeatures
+        &preparedTrees->BinFeatureCombinationClass,
+        &preparedTrees->CombinationClassFeatures
     );
+}
 
-    TProfileInfo processTreesProfile(treeCount);
+static void InitLeafWeights(
+    const TFullModel& model,
+    const TDataProvider* dataset,
+    NPar::TLocalExecutor* localExecutor,
+    TVector<double>* leafWeights
+) {
+    const auto& leafWeightsOfModels = model.ModelTrees->GetLeafWeights();
+    if (leafWeightsOfModels.empty()) {
+        CB_ENSURE(
+                dataset,
+                "To calculate shap values, either a model with leaf weights, or a dataset are required."
+        );
+        CB_ENSURE(dataset->ObjectsGrouping->GetObjectCount() != 0, "no docs in pool");
+        CB_ENSURE(dataset->MetaInfo.GetFeatureCount() > 0, "no features in pool");
```

Comment:
To calculate shap values, dataset must contain objects.

To calculate shap values, dataset must contain features.